File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/scope_functions/java/ScopeFunctionsExample.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.scope_functions.java;

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.java.User;

import java.util.List;

public class ScopeFunctionsExample {

    public User createAndPrintUser() {
        User user = new User(
                "John",
                "Doe",
                30,
                List.of("USER")
        );

        // Java usually uses explicit imperative steps.
        System.out.println("Created user: " + user.getFullName());

        int nameLength = user.getFullName().length();

        System.out.println("Full name length: " + nameLength);

        return user;
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/scope_functions/kotlin/ScopeFunctionsExample.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.scope_functions.kotlin

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.kotlin.User

class ScopeFunctionsExample {

    fun createAndPrintUser(): User {
        val user = User(
            firstName = "John",
            lastName = "Doe",
            age = 30,
            roles = listOf("USER")
        ).also {
            // also is useful for side effects,
            // such as logging or debugging.
            println("Created user: ${it.fullName}")
        }

        // let transforms the object into another value.
        val nameLength = user.fullName.let {
            it.length
        }

        println("Full name length: $nameLength")

        return user
    }
}